Place all invocations of read on stream objects into try-catch blocks

xiaofan-cerebras opened the pull request:
All invocations of the read function on stream objects can throw "insufficient data" exceptions. They should be placed inside try-catch blocks in order to achieve maximum possible parsing depth.

xiaofan-cerebras committed:
All invocations of the read function on stream objects can throw "insufficient data" exceptions. They should be placed inside try-catch blocks in order to achieve maximum possible parsing depth.

ebertland commented:
Which case(s) actually caused an uncaught exception or situation where `malformed(true)` was not called?

xiaofan-cerebras commented:
The original problem happened to a packet with a corrupted IPv4 "total length" field. The IP packet carries an ICMP payload of the type "TIMESTAMP_REPLY". This type of ICMP requires an additional 8 bytes of data after the ICMP header, and this part of the parsing was outside the try-catch block, leading to an uncaught exception.

However, after reviewing the code, I feel that this kind of uncaught exceptions can happen in other places where the read function is called outside of try-catch blocks, such as IPv4 options parsing.

xiaofan-cerebras committed:
- Add a child class of the InputMemoryStream to handle insufficient_data exceptions.
- Update the header classes to use the child class.
- Remove the try-catch blocks put in place to handle the insufficient_data exceptions.

xiaofan-cerebras commented:
I've pushed a second commit to roll back the per-header try-catch blocks and use a child class of InputMemeoryStream class to handle the insufficient_data exception.

xiaofan-cerebras commented:
The `read` function of the InputMemoryStream class is a template function. Since C++ doesn't allow virtual template functions, I chose to target the `can_read` function instead.

ebertland commented on include/tins/memory_helpers.h:
```diff
@@ -113,9 +114,17 @@ class InputMemoryStream {
+    virtual bool check_can_read(size_t byte_count) const {
+        if (!can_read(byte_count)) {
             throw insufficient_data();
+            return false;
```
This return is unreachable. Did you add it because of a warning or other check?

xiaofan-cerebras replied:
I was worried about a compiler warning for not returning on the conditional, but apparently the compiler is happy with not having the return here, so I've pushed a change to remove it.

ebertland commented on include/tins/memory_helpers.h:
```diff
@@ -141,11 +150,40 @@ class InputMemoryStream {
     void read(IPv4Address& address);
     void read(IPv6Address& address);
-private:
+protected:
```
I don't see either of these data members used in the PduInputMemoryStream derived class.

xiaofan-cerebras replied:
I originally initialized these two members in the child class constructor, which requires that these members be protected instead of private. I've resolved this issue by calling the parent class constructor in the child class constructor.

ebertland commented on src/icmp.cpp:
```diff
@@ -78,6 +73,7 @@ ICMP::ICMP(const uint8_t* buffer, uint32_t total_sz)
     }
     // Attempt to parse ICMP extensions
     try_parse_extensions(stream);
+
```
nit: Unnecessary blank line.

ebertland commented on src/ip.cpp:
```diff
@@ -130,6 +125,7 @@ IP::IP(const uint8_t* buffer, uint32_t total_sz) {
             options_.push_back(option(opt_type));
         }
     }
+
```
nit: Unnecessary blank line.

(The author doesn't use much whitespace (or any?) within functions. I'm not a big fan of that style, but we should try to follow his style.)

ebertland commented on src/tcp.cpp:
```diff
@@ -125,6 +120,7 @@ TCP::TCP(const uint8_t* buffer, uint32_t total_sz) {
             stream.skip(len);
         }
     }
+
```
nit: Unnecessary whitespace.

ebertland reviewed (changes requested):
Looks good. Just a few minor changes requested.

ebertland commented:
Did you run the test suite?

ebertland commented:
There are a few places where you change the function argument type to from InputMemoryStream to PduInputMemoryStream where it's probably not necessary. In other words, just allow implicit cast to the base class. Since the functions you overrode are virtual, it should work fine. The only reason to change that would be to reduce the size of the change if we ever try to upstream this enhancement. I'm happy with the way you did it, though.

xiaofan-cerebras committed:
Adjustments based on suggestions in PR5.

xiaofan-cerebras commented:
> There are a few places where you change the function argument type to from InputMemoryStream to PduInputMemoryStream where it's probably not necessary. In other words, just allow implicit cast to the base class. Since the functions you overrode are virtual, it should work fine. The only reason to change that would be to reduce the size of the change if we ever try to upstream this enhancement. I'm happy with the way you did it, though.

I think it's a good idea to use the parent class reference instead of the child class one, this minimizes the code changes and gives the callers of the functions more flexibility.

I've pushed a commit to make this change.

xiaofan-cerebras commented:
> Did you run the test suite?

I reran the initial failed simulation using this build of the library, and the simulation is able to pass. Running the simulation with the normal libtins build reproduced the failure. So these changes seem to fix the problem.

ebertland commented:
> > Did you run the test suite?
> 
> I reran the initial failed simulation using this build of the library, and the simulation is able to pass. Running the simulation with the normal libtins build reproduced the failure. So these changes seem to fix the problem.

I mean the unit tests that come with libtins. See `Running tests` in README.md.

xiaofan-cerebras commented:
> > > Did you run the test suite?
> > 
> > 
> > I reran the initial failed simulation using this build of the library, and the simulation is able to pass. Running the simulation with the normal libtins build reproduced the failure. So these changes seem to fix the problem.
> 
> I mean the unit tests that come with libtins. See `Running tests` in README.md.

I have run the unit tests with the changes. All tests passed.

ebertland commented:
Looks good. You should do a "squash and merge" for this change.
